### DISCUSSION TITLE
pico W not finding uPip

### DISCUSSION THREAD

**Discussion by JohnnyTank** [micropython/micropython#10546] [RP2040 / Pico] [answered]
I tried to install Phew! via upip.

```
import wlan

wlan.do_connect()
import upip
upip.install('micropython-phew')
```
I have one board whichis installing the package properly and at least one board is dropping an Error - missing uPip-Module.
I am using the latest Firmware-Version:  v1.19.1-831-g4f3780a15 on 2023-01-20.
The working board is using an older Version: MicroPython v1.19.1 on 2022-08-31; Raspberry Pi Pico W with RP2040

**Comment by scruss** [comment, marked as answer]
upip is no longer supported, and has been replaced (mostly) by *mip*. See [upip to mip migration](https://github.com/orgs/micropython/discussions/9565) for a little more clarity.

**Comment by jimmo** [comment]
@Gadgetoid Would you be willing to add a package.json for phew?

**Reply by Gadgetoid** [reply]
Sure- I think we have another library that might need the same treatment. I'll look into this now.

**Reply by Gadgetoid** [reply]
I have a question regarding version pinning which relates to `phew` too - https://github.com/pimoroni/pms5003-micropython/pull/3

Glancing over the sourcecode leads me to think it's the latter, but does `mip` attempt to pull a specific tagged release from GitHub, or just the latest, potentially "unstable", files?

Forgive me if I'm missed something above, in the migration note or the docs- my brain is cold-addled fuzz today.

**Reply by jimmo** [reply]
@Gadgetoid If the user does `mip.install('github:foo/bar')` then we pull from your default branch. If they do `mip.install('github:foo/bar', version='dev')` then we pull from the 'dev' tag or branch (tags and branches share the same namespace, so this won't be ambiguous).

(Same applies to `mpremote mip install --version=... github:foo/bar` ).

**Reply by jimmo** [reply]
One extra piece of context that is probably helpful is that the version passed to mip.install is what's used to get the package.json, but then this version is also passed to subsequent entries in the `files` mentioned in the package.json.

So if you do

`mip.install('github:pimoroni/phew`, version='branch-or-tag-name')` then that will fetch `https://raw.githubusercontent.com/pimoroni/phew/branch-or-tag-name/package.json`, which will then say to fetch e.g. `github:pimoroni/phew/phew/__init__.py` but then this is implicitly turned into `https://raw.githubusercontent.com/pimoroni/phew/branch-or-tag-name/phew/__init__.py`.

So there's a bit of magic here, but I think it's the entirely sensible and expected thing to do.

**Reply by Gadgetoid** [reply]
Thank you! I don't expect us to make much first party use of this, but I think it's very important to understand all the moving parts. And it wouldn't kill us to write more *python* packages.

`mip.install('github:pimoroni/phew')` is working beautifully.
